lib/Sema/CSGen.cpp

```diff
@@ -1916,61 +1917,39 @@ namespace {
           options |= TypeResolutionFlags::AllowPackReferences;
           elementEnv = OuterExpansions.back();
         }
-        const auto result = TypeResolution::resolveContextualType(
+        auto result = TypeResolution::resolveContextualType(
             specializationArg, CurDC, options,
             // Introduce type variables for unbound generics.
             OpenUnboundGenericType(CS, locator),
             HandlePlaceholderType(CS, locator),
             OpenPackElementType(CS, locator, elementEnv));
-        if (result->hasError())
-          return true;
-
+        if (result->hasError()) {
+          auto &ctxt = CS.getASTContext();
+          auto *repr = new (ctxt) PlaceholderTypeRepr(specializationArg->getLoc());
+          result = PlaceholderType::get(ctxt, repr);
+          hasError = true;
+        }
         specializationArgTypes.push_back(result);
       }
 
       CS.addConstraint(
           ConstraintKind::ExplicitGenericArguments, boundType,
           PackType::get(CS.getASTContext(), specializationArgTypes),
           locator);
-      return false;
+      return hasError;
     }
 
     Type visitUnresolvedSpecializeExpr(UnresolvedSpecializeExpr *expr) {
       auto baseTy = CS.getType(expr->getSubExpr());
-
-      if (baseTy->isTypeVariableOrMember()) {
-        return baseTy;
-      }
-
-      // We currently only support explicit specialization of generic types.
-      // FIXME: We could support explicit function specialization.
-      auto &de = CS.getASTContext().Diags;
-      if (baseTy->is<AnyFunctionType>()) {
-        de.diagnose(expr->getSubExpr()->getLoc(),
-                    diag::cannot_explicitly_specialize_generic_function);
+      auto *overloadLocator = CS.getConstraintLocator(expr->getSubExpr());
+      if (addSpecializationConstraint(overloadLocator, baseTy->getMetatypeInstanceType(),
+                                      expr->getUnresolvedParams())) {
+        // Diagnosis of failed type lookup - add a bit of context
+        auto &de = CS.getASTContext().Diags;
```

Comment:
I still think we should delay simplification here to let "normal" solving fail and re-attempt during diagnostic mode.

Reply:
Sorry, I think you'll have to be more explicit, I'm not sure I understand.

`CS.addConstraint()` happens inside of `addSpecializationConstraint()`. If this is changed to be `CS.addUnsolvedConstraint()`, then the solve for many of these cases will type check while leaving the specialization constraint in the Inactive list the whole time (because they really are extraneous), and then fail due to remaining inactive constraints (CSStep.cpp:482).

The if statement here was just for whether or not to add a note, the constraint is added no matter what. But I'll clean that part up.

Reply:
To be precise my suggestion was to remove the result type for `addSpecializationConstraint`, inside of that method: create a new `ExplicitGenericArguments` constraint, use `addUnsolvedConstraint` + `activateConstraint` (usually `activateConstraint` is not necessary but as far as I understand the main problem here that sometimes there are no type variables here for solver to bind and subsequently re-evaluate the constraint).

One trick to avoid explicit activation would be to allocate a new type variable and use an unsolved `Equal` constraint to connect it to the `boundType`, type variable itself should be used in place of `boundType` in an unsolved `ExplicitGenericArguments` constraint. This way the solver would pick the type variable and bind it during solving which activates all of the constraints that reference it.

Reply:
Gotcha. Thanks for expanding.